src/model/TexImage.cpp

```diff
@@ -66,93 +50,48 @@ auto TexImage::cairoReadFunction(TexImage* image, unsigned char* data, unsigned
     return CAIRO_STATUS_SUCCESS;
 }
 
-/**
- * Sets the binary data, a .PNG image or a .PDF
- */
-void TexImage::setBinaryData(string binaryData) { this->binaryData = std::move(binaryData); }
-
 /**
  * Gets the binary data, a .PNG image or a .PDF
  */
-auto TexImage::getBinaryData() -> string& { return this->binaryData; }
+auto TexImage::getBinaryData() const -> std::string const& { return this->binaryData; }
 
 void TexImage::setText(string text) { this->text = std::move(text); }
 
 auto TexImage::getText() -> string { return this->text; }
 
-auto TexImage::getImage() -> cairo_surface_t* {
-    if (this->image == nullptr && !this->parsedBinaryData) {
-        loadBinaryData();
-    }
-
-    return this->image;
-}
-
-/**
- * Load the binary data, either .PNG or .PDF
- */
-void TexImage::loadBinaryData() {
-    freeImageAndPdf();
-
+auto TexImage::loadData(std::string&& bytes, GError** err) -> bool {
+    this->freeImageAndPdf();
+    this->binaryData = bytes;
     if (this->binaryData.length() < 4) {
-        this->parsedBinaryData = true;
-        return;
+        return false;
     }
 
-    string type = this->binaryData.substr(0, 4);
-
-    if (type[1] == 'P' && type[2] == 'N' && type[3] == 'G') {
-        this->read = 0;
+    const std::string type = binaryData.substr(1, 3);
+    if (type == "PDF") {
+        // Note: binaryData must not be modified while pdf is live.
+        this->pdf = poppler_document_new_from_data(this->binaryData.data(), this->binaryData.size(), nullptr, err);
+        if (!pdf || poppler_document_get_n_pages(this->pdf) < 1) {
+            return false;
+        }
+        PopplerPage* page = poppler_document_get_page(this->pdf, 0);
+        poppler_page_get_size(page, &this->width, &this->height);
```

Comment:
The width and height should only be set for newly created TexImages. If the TexImage is loaded from a file, the saved width and height should not be overwritten.